File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.cpp
class Solution {
public:
    int minimumPushes(string word) {
        
        vector<int>arr;
        unordered_map<char,int>hashmap;
        for (char c:word)
        {
            hashmap[c]++;

        }
       for(auto &val:hashmap)
       {
        arr.push_back(val.second);
       }
       
       sort(arr.begin(),arr.end(),greater<int>());
       int ans=0;
       for(int i=0;i<arr.size();i++)
       {
         std::cout << arr[i];
         ans+=arr[i]*(1+ (i/8));
       }
    
       return ans;
    }
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.java
class Solution {
    public int minimumPushes(String word) {
        Map<String, Integer> hm = new HashMap<>();
        if (word.length() <= 8) {
            return word.length();
        }
        for (int i = 0; i < word.length(); i++) {
            hm.put(String.valueOf(word.charAt(i)), hm.getOrDefault(String.valueOf(word.charAt(i)),0)+1);
        }

        PriorityQueue<Map.Entry<String,Integer>> pq = new PriorityQueue<>(
            Map.Entry.<String, Integer>comparingByValue().reversed()
        );
        pq.addAll(hm.entrySet());
        int count = 0;
        int distinct = 0;
        while (!pq.isEmpty()) {
            Map.Entry<String, Integer> entry = pq.poll();
            int value = entry.getValue();
            //System.out.println("count: "+count+"key: "+entry.getKey()+"value: "+entry.getValue());
            for (int i  = 0; i < value; i++) {
                count = count + 1+ distinct/8;
            }
            distinct++;
        }
        return count;
    }
}